java/src/org/openqa/selenium/manager/SeleniumManager.java

```diff
@@ -117,7 +117,7 @@ private static String runCommand(String... command) {
    * Determines the correct Selenium Manager binary to use.
    * @return the path to the Selenium Manager binary.
    */
-    private File getBinary() {
+    private synchronized File getBinary() {
```

Comment:
@bonigarcia - This would cause all threads to go through the process of `obtain lock, check binary variable, release lock`. We could perhaps instead consider implement a doubly checked locking mechanism ?

Something like this ?

```java
  private File getBinary() {
    if (binary != null) {
      return binary;
    }
    synchronized (this) {
      if (binary != null) {
        return binary;
      }
      try {
        Platform current = Platform.getCurrent();
        String folder = "linux";
        String extension = "";
        if (current.is(WINDOWS)) {
          extension = EXE;
          folder = "windows";
        } else if (current.is(MAC)) {
          folder = "macos";
        }
        String binaryPath = String.format("%s/%s%s", folder, SELENIUM_MANAGER, extension);
        try (InputStream inputStream = this.getClass().getResourceAsStream(binaryPath)) {
          Path tmpPath = Files.createTempDirectory(SELENIUM_MANAGER + System.nanoTime());
          File tmpFolder = tmpPath.toFile();
          tmpFolder.deleteOnExit();
          binary = new File(tmpFolder, SELENIUM_MANAGER + extension);
          Files.copy(inputStream, binary.toPath(), REPLACE_EXISTING);
        }
        binary.setExecutable(true);
      } catch (Exception e) {
        throw new WebDriverException("Unable to obtain Selenium Manager", e);
      }
    }
    return binary;
  }
```

Reply:
Synchronization at the method level will do the trick as well.

Reply:
No. Its different. Method level synchronisation will cause threads to obtain lock to enter into the `getBinary()` method. A doubly checked lock backed implementation will ensure that only those threads that are going to initialise the value of `binary` will take the lock. All the subsequent threads will return without having to take the lock.

Reply:
I agree it's different, but the difference is minimum, IMO. What I mean with my comment is that the corrurency problems detected are solved with the synchronization at the method level.